File: python/interview_34.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        if root is None:
            return []
        result = []
        item = []
        def dfs(node, target):
            if node.left is None and node.right is None:
                if node.val == target:
                    item.append(node.val)
                    result.append(item[:])
                    item.pop()
                return
            item.append(node.val)
            if node.left is not None:
                dfs(node.left, target - node.val)
            if node.right is not None:
                dfs(node.right, target - node.val)
            item.pop()
        dfs(root, sum)
        return result

File: Java/interview_34.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    List<List<Integer>> result;
    LinkedList<Integer> item;

    private void visit(TreeNode node, int target) {
        if (node.left == null && node.right == null) {
            if (target == node.val) {
                item.add(node.val);
                result.add(new LinkedList<Integer>(item));
                item.removeLast();
            }
            return;
        }
        item.add(node.val);
        if (node.left != null) {
            visit(node.left, target - node.val);
        }
        if (node.right != null) {
            visit(node.right, target - node.val);
        }
        item.removeLast();
    }

    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        result = new LinkedList<>();
        if (root == null) {
            return result;
        }
        item = new LinkedList<>();
        visit(root, sum);
        return result;
    }
}
